File: 2022/java/Day9.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

class Day9 {

    public static void main(String ... args) throws Exception {
        var moves = parseMoves(Files.readString(Path.of("../input/day9.txt")));
        var head = tieKnots(10);
        head = runMoves(moves, head);
        System.out.printf("part 1: %s\n", nth(head, 1).history.stream().distinct().count());
        System.out.printf("part 2: %s\n", nth(head, 9).history.stream().distinct().count());
    }

    static Knot nth(Knot k, int count) {
        return count == 0 ? k : nth(k.tail, count-1);
    }

    static Knot runMoves(List<Move> moves, Knot head) {
        return moves.stream().reduce(head, (k, m) -> k.apply(m), (a,b) -> {return a;});
    }

    static Knot tieKnots(int count) {
        return new Knot(new XY(0,0), Set.of(new XY(0,0)), count > 1 ? tieKnots(count-1): null);
    }

    static List<Move> parseMoves(String input) {
        return input.lines()
            .map(l -> new Move(vector.get(l.charAt(0)), Integer.parseInt(l.substring(2))))
            .toList();
    }

    record Knot(XY pos, Set<XY> history, Knot tail) {
        Knot apply(Move m) {
            return IntStream.range(0, m.steps).mapToObj(i -> m)
                .reduce(this, (k, move) -> k.move(move.vector), (a,b) -> {return a;});
        }

        Knot move(XY vector) {
            var newPos = pos.move(vector);
            var newHistory = Stream.concat(history.stream(), Stream.of(newPos)).collect(Collectors.toSet());
            return new Knot(newPos, newHistory, tail != null ? tail.follow(newPos) : null);
        }

        Knot follow(XY head) {
            if (head.touching(pos)) {
                return this;
            }
            var dx = (head.x-pos.x)/2 + (head.x-pos.x)%2;
            var dy = (head.y-pos.y)/2 + (head.y-pos.y)%2;
            return move(new XY(dx, dy));
        }
    }

    record Move(XY vector, int steps) {}
    
    record XY(int x, int y) {
        XY move(XY vector) {
            return new XY(x + vector.x, y + vector.y);
        }
        boolean touching(XY o) {
            return Math.abs(x-o.x) <= 1 && Math.abs(y-o.y) <= 1;
        }
    }
    
    static Map<Character,XY> vector = Map.of(
        'U', new XY(0, -1), 'D', new XY(0, +1),
        'L', new XY(-1, 0), 'R', new XY(+1, 0));
}

File: 2022/go/day9/day9.go
package main

import (
	"fmt"
	"math"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day9.txt")
	moves := parseMoves(string(input))
	head := tieKnots(10)
	head = runMoves(head, moves)
	fmt.Printf("part1 : %d\n", len(head.getNth(1).visited))
	fmt.Printf("part2 : %d\n", len(head.getNth(9).visited))
}

func tieKnots(count int) *knot {
	knot := newKnot()
	prev := knot
	for i := 1; i < count; i++ {
		prev.tail = newKnot()
		prev = prev.tail
	}
	return knot
}

func runMoves(knot *knot, moves []move) *knot {
	for _, m := range moves {
		for s := byte(0); s < m.steps; s++ {
			knot.move(dirs[m.dir])
		}
	}
	return knot
}

type pos struct {
	x, y int
}

type knot struct {
	pos     pos
	visited map[pos]bool
	tail    *knot
}

type move struct {
	dir, steps byte
}

func (k *knot) getNth(count int) *knot {
	if count == 0 {
		return k
	}
	return k.tail.getNth(count - 1)
}

func (k *knot) move(v pos) {
	k.pos = k.pos.move(v)
	k.visited[k.pos] = true
	if k.tail != nil {
		k.tail.follow(k)
	}
}

func (p pos) move(v pos) pos {
	return pos{p.x + v.x, p.y + v.y}
}

func (tail *knot) follow(head *knot) {
	if tail.pos.touching(head.pos) {
		return
	}
	dx := (head.pos.x-tail.pos.x)/2 + (head.pos.x-tail.pos.x)%2
	dy := (head.pos.y-tail.pos.y)/2 + (head.pos.y-tail.pos.y)%2
	tail.move(pos{dx, dy})
}

func (k1 pos) touching(k2 pos) bool {
	return math.Abs(float64(k1.x)-float64(k2.x)) < 1.5 && math.Abs(float64(k1.y)-float64(k2.y)) < 1.5
}

func parseMoves(input string) (moves []move) {
	lines := strings.Split(input, "\n")
	for _, line := range lines {
		var m move
		fmt.Sscanf(line, "%c %d", &m.dir, &m.steps)
		moves = append(moves, m)
	}
	return moves
}

func newKnot() *knot {
	k := &knot{visited: make(map[pos]bool)}
	k.move(pos{})
	return k
}

var dirs = map[byte]pos{'U': {0, -1}, 'D': {0, 1}, 'L': {-1, 0}, 'R': {1, 0}}
